File: src/team_agent/messaging/scheduler.py
from __future__ import annotations

from team_agent.messaging.deps import (
    EventLog,
    MessageStore,
    check_team_owner,
    datetime,
    json,
    load_runtime_state,
    load_spec,
    save_runtime_state,
    send_message,
    timedelta,
    timezone,
)

from pathlib import Path
from typing import Any

_ACTIVE_TASK_STATUSES = {"pending", "assigned", "in_progress", "ready", "running", "needs_retry"}
_INBOUND_WORK_STATUSES = {"pending", "accepted", "target_resolved", "injected"}
_DELIVERED_MESSAGE_STATUSES = {"visible", "submitted", "delivered", "acknowledged"}
_PROGRESS_EVENTS = {
    "mcp.report_result",
    "report_result.accepted",
    "send.deliver_attempt",
    "send.submitted",
    "leader_receiver.deliver_attempt",
    "leader_receiver.submitted",
    "communication.peer_mirrored",
}
_RESTART_RESET_EVENTS = {"restart.agent_start", "restart.complete", "reset_agent.complete", "start_agent.complete"}
_ALERT_TYPES = {"stuck", "idle_fallback"}

def _fire_due_scheduled_events(workspace: Path, store: MessageStore, event_log: EventLog) -> list[int]:
    fired: list[int] = []
    for row in store.due_scheduled_events():
        payload = json.loads(row["payload_json"] or "{}")
        try:
            if row["kind"] == "send":
                result = send_message(
                    workspace,
                    row["target"],
                    str(payload.get("content") or ""),
                    task_id=payload.get("task_id"),
                    sender=payload.get("sender", "coordinator"),
                    requires_ack=bool(payload.get("requires_ack", True)),
                    wait_visible=bool(payload.get("wait_visible", True)),
                    timeout=float(payload.get("timeout", 30)),
                )
            elif row["kind"] == "health_ping":
                result = {"ok": True, "status": "logged"}
                event_log.write("coordinator.health_ping", target=row["target"], payload=payload)
            else:
                result = {"ok": False, "error": f"unknown scheduled event kind: {row['kind']}"}
            if not result.get("ok") and row["kind"] == "send":
                retry = _schedule_send_retry(store, row, payload, result)
                if retry:
                    result = {**result, **retry}
                    store.mark_scheduled_event(int(row["id"]), "retry_scheduled", result)
                    event_log.write(
                        "coordinator.scheduled_retry",
                        id=row["id"],
                        retry_event_id=retry["retry_event_id"],
                        target=row["target"],
                        attempt=retry["next_attempt"],
                    )
                    fired.append(int(row["id"]))
                    continue
            store.mark_scheduled_event(int(row["id"]), "done" if result.get("ok") else "failed", result)
            fired.append(int(row["id"]))
        except Exception as exc:
            result = {"ok": False, "error": str(exc)}
            store.mark_scheduled_event(int(row["id"]), "failed", result)
            event_log.write("coordinator.scheduled_failed", id=row["id"], error=str(exc))
    return fired


def _schedule_send_retry(
    store: MessageStore,
    row: dict[str, Any],
    payload: dict[str, Any],
    result: dict[str, Any],
) -> dict[str, Any] | None:
    attempt = int(payload.get("attempt") or 1)
    max_attempts = int(payload.get("max_attempts") or 1)
    if attempt >= max_attempts:
        return None
    retry_payload = dict(payload)
    retry_payload["attempt"] = attempt + 1
    due_at = datetime.now(timezone.utc) + timedelta(seconds=min(2 * attempt, 5))
    retry_id = store.add_scheduled_event(due_at.isoformat(), row["target"], row["kind"], retry_payload)
    return {
        "retry_event_id": retry_id,
        "next_attempt": attempt + 1,
        "max_attempts": max_attempts,
        "retry_reason": result.get("reason") or result.get("error"),
    }


def _detect_stuck_agents(
    workspace: Path,
    state: dict[str, Any],
    store: MessageStore,
    event_log: EventLog,
) -> list[str]:
    spec_path = Path(state.get("spec_path", workspace / "team.spec.yaml"))
    spec = load_spec(spec_path) if spec_path.exists() else {}
    runtime_cfg = spec.get("runtime", {})
    stuck_timeout = int(runtime_cfg.get("stuck_timeout_sec", 300))
    push_min_interval = int(runtime_cfg.get("push_min_interval_sec", 60))
    health = store.agent_health()
    stuck: list[str] = []
    now = datetime.now(timezone.utc)
    for agent_id, row in health.items():
        if row.get("status") not in {"RUNNING"} or not row.get("last_output_at"):
            continue
        try:
            last = datetime.fromisoformat(row["last_output_at"])
        except ValueError:
            continue
        if last.tzinfo is None:
            last = last.replace(tzinfo=timezone.utc)
        if (now - last).total_seconds() < stuck_timeout:
            continue
        suppression = _active_alert_suppression(state, store, event_log, agent_id, "stuck")
        has_work, work_reason = _agent_has_stuck_relevant_work(state, store, agent_id)
        if not has_work:
            event_log.write("coordinator.agent_stuck_suppressed", agent_id=agent_id, reason="idle_no_work", last_output_at=row["last_output_at"])
            continue
        if suppression:
            continue
        progress_event = _recent_agent_progress_event(event_log, agent_id, last)
        if progress_event:
            event_log.write(
                "coordinator.agent_stuck_suppressed",
                agent_id=agent_id,
                reason="recent_progress_event",
                progress_event=progress_event.get("event"),
                progress_ts=progress_event.get("ts"),
                last_output_at=row["last_output_at"],
                work_reason=work_reason,
            )
            continue
        stuck.append(agent_id)
        state.setdefault("coordinator", {})
        push_key = f"last_stuck_push_at:{agent_id}"
        last_push_raw = state["coordinator"].get(push_key)
        should_push = True
        if last_push_raw:
            try:
                last_push = datetime.fromisoformat(last_push_raw)
                if last_push.tzinfo is None:
                    last_push = last_push.replace(tzinfo=timezone.utc)
                should_push = (now - last_push).total_seconds() >= push_min_interval
            except ValueError:
                should_push = True
        event_log.write("coordinator.agent_stuck", agent_id=agent_id, last_output_at=row["last_output_at"], work_reason=work_reason)
        if should_push:
            state["coordinator"][push_key] = now.isoformat()
            try:
                send_message(
                    workspace,
                    "leader",
                    f"agent {agent_id} appears stuck: no output for {stuck_timeout}s",
                    sender="coordinator",
                    requires_ack=False,
                    wait_visible=False,
                )
            except Exception as exc:
                event_log.write("coordinator.stuck_push_failed", agent_id=agent_id, error=str(exc))
    return stuck


def stuck_list(workspace: Path) -> dict[str, Any]:
    state = load_runtime_state(workspace)
    suppressed = state.get("coordinator", {}).get("suppressed_idle_alerts", {})
    return {"ok": True, "suppressed_idle_alerts": suppressed}


def stuck_cancel(
    workspace: Path,
    agent_id: str,
    alert_type: str = "stuck",
    suppressed_by: str = "leader",
) -> dict[str, Any]:
    if alert_type == "all":
        alert_types = sorted(_ALERT_TYPES)
    elif alert_type in _ALERT_TYPES:
        alert_types = [alert_type]
    else:
        return {"ok": False, "status": "refused", "reason": "invalid_alert_type", "alert_type": alert_type}
    state = load_runtime_state(workspace)
    gate = check_team_owner(state)
    if gate:
        return gate
    store = MessageStore(workspace)
    coordinator = state.setdefault("coordinator", {})
    suppressed = coordinator.setdefault("suppressed_idle_alerts", {})
    agent_suppressions = suppressed.setdefault(agent_id, {})
    now = datetime.now(timezone.utc).isoformat()
    snapshot = _agent_alert_snapshot(state, store, agent_id)
    for item in alert_types:
        agent_suppressions[item] = {
            "suppressed_at": now,
            "suppressed_by": suppressed_by,
            "snapshot": snapshot,
        }
    save_runtime_state(workspace, state)
    EventLog(workspace).write("coordinator.idle_alert_suppressed", agent_id=agent_id, alert_types=alert_types, suppressed_by=suppressed_by)
    return {"ok": True, "agent_id": agent_id, "alert_types": alert_types, "suppressed": agent_suppressions}


def _active_alert_suppression(
    state: dict[str, Any],
    store: MessageStore,
    event_log: EventLog,
    agent_id: str,
    alert_type: str,
) -> dict[str, Any] | None:
    entry = state.get("coordinator", {}).get("suppressed_idle_alerts", {}).get(agent_id, {}).get(alert_type)
    if not isinstance(entry, dict):
        return None
    cleared = _suppression_clear_reason(state, store, event_log, agent_id, entry)
    if cleared:
        _clear_alert_suppression(state, agent_id, alert_type)
        event_log.write("coordinator.idle_alert_suppression_cleared", agent_id=agent_id, alert_type=alert_type, reason=cleared)
        return None
    return entry


def _suppression_clear_reason(
    state: dict[str, Any],
    store: MessageStore,
    event_log: EventLog,
    agent_id: str,
    entry: dict[str, Any],
) -> str | None:
    previous = entry.get("snapshot") if isinstance(entry.get("snapshot"), dict) else {}
    current = _agent_alert_snapshot(state, store, agent_id)
    if current.get("assigned_task_ids") != previous.get("assigned_task_ids"):
        return "task_assignment_changed"
    if current.get("delivered_message_ids") != previous.get("delivered_message_ids"):
        return "inbound_delivery_changed"
    try:
        suppressed_at = datetime.fromisoformat(str(entry.get("suppressed_at")))
    except ValueError:
        return "invalid_suppression_timestamp"
    if suppressed_at.tzinfo is None:
        suppressed_at = suppressed_at.replace(tzinfo=timezone.utc)
    if _recent_agent_progress_event(event_log, agent_id, suppressed_at):
        return "progress_event"
    if _recent_restart_or_reset_event(event_log, agent_id, suppressed_at):
        return "restart_or_reset"
    return None


def _clear_alert_suppression(state: dict[str, Any], agent_id: str, alert_type: str) -> None:
    suppressed = state.get("coordinator", {}).get("suppressed_idle_alerts", {})
    agent_suppressions = suppressed.get(agent_id, {})
    agent_suppressions.pop(alert_type, None)
    if not agent_suppressions:
        suppressed.pop(agent_id, None)


def _agent_alert_snapshot(state: dict[str, Any], store: MessageStore, agent_id: str) -> dict[str, Any]:
    assigned_task_ids = sorted(str(task.get("id")) for task in state.get("tasks", []) if task.get("assignee") == agent_id)
    delivered_message_ids = sorted(
        str(message.get("message_id"))
        for message in store.messages()
        if message.get("recipient") == agent_id and message.get("status") in _DELIVERED_MESSAGE_STATUSES
    )
    return {"assigned_task_ids": assigned_task_ids, "delivered_message_ids": delivered_message_ids}


def _agent_has_stuck_relevant_work(state: dict[str, Any], store: MessageStore, agent_id: str) -> tuple[bool, str]:
    for task in state.get("tasks", []):
        if task.get("assignee") == agent_id and task.get("status", "pending") in _ACTIVE_TASK_STATUSES:
            return True, "active_task"
    for message in store.messages():
        if message.get("recipient") == agent_id and message.get("status") in _INBOUND_WORK_STATUSES:
            return True, "inbound_message"
    return False, "idle_no_work"


def _recent_agent_progress_event(event_log: EventLog, agent_id: str, since: datetime) -> dict[str, Any] | None:
    for event in reversed(event_log.tail(200)):
        if event.get("event") not in _PROGRESS_EVENTS:
            continue
        if not _event_mentions_agent(event, agent_id):
            continue
        try:
            ts = datetime.fromisoformat(str(event.get("ts")))
        except ValueError:
            continue
        if ts.tzinfo is None:
            ts = ts.replace(tzinfo=timezone.utc)
        if ts >= since:
            return event
    return None


def _event_mentions_agent(event: dict[str, Any], agent_id: str) -> bool:
    if event.get("agent_id") == agent_id or event.get("sender") == agent_id or event.get("target") == agent_id:
        return True
    payload = event.get("payload")
    return isinstance(payload, dict) and (payload.get("from") == agent_id or payload.get("to") == agent_id)


def _recent_restart_or_reset_event(event_log: EventLog, agent_id: str, since: datetime) -> dict[str, Any] | None:
    for event in reversed(event_log.tail(200)):
        if event.get("event") not in _RESTART_RESET_EVENTS:
            continue
        if event.get("agent_id") != agent_id and agent_id not in set(event.get("agents") or []):
            continue
        try:
            ts = datetime.fromisoformat(str(event.get("ts")))
        except ValueError:
            continue
        if ts.tzinfo is None:
            ts = ts.replace(tzinfo=timezone.utc)
        if ts >= since:
            return event
    return None

File: crates/team-agent/src/messaging/scheduler.rs
//! scheduler.py — 调度器心脏 + stuck 检测 + 告警抑制 (card §18/§68/§73)。

use std::path::Path;

use rusqlite::params;

use crate::event_log::EventLog;
use crate::message_store::MessageStore;
use crate::transport::{PaneId, Transport};

use super::delivery::{deliver_stored_message, handle_trust_retry_needed};
use super::helpers::{parse_scheduled_kind, status_wire};
use super::{AlertType, MessagingError, ScheduledKind, TrustRetryPayload, TRUST_RETRY_MAX_ATTEMPTS};

/// `_fire_due_scheduled_events` (`scheduler.py:41`):coordinator tick 的调度器心脏 —— 分派到期
/// `send`/`health_ping`/`trust_retry` ([`ScheduledKind`] 穷尽 match),send 失败有界重试。
/// 返回 fired 的 scheduled_event id 列表。**daemon-path** (step 12 调) → Result。
pub fn fire_due_scheduled_events(
    workspace: &Path,
    store: &MessageStore,
    transport: &dyn Transport,
    event_log: &EventLog,
) -> Result<Vec<i64>, MessagingError> {
    let _ = transport;
    let conn = crate::db::schema::open_db(store.db_path())?;
    let due_events = {
        let mut stmt = conn.prepare(
            "select id, kind, target, payload_json from scheduled_events where status = 'pending' and due_at <= ?1 order by due_at, id",
        )?;
        let rows = stmt.query_map(params![chrono::Utc::now().to_rfc3339()], |row| {
            Ok((
                row.get::<_, i64>(0)?,
                row.get::<_, String>(1)?,
                row.get::<_, String>(2)?,
                row.get::<_, String>(3)?,
            ))
        })?;
        rows.collect::<Result<Vec<_>, _>>()?
    };
    let mut fired = Vec::new();
    for (id, kind, target, payload_json) in due_events {
        let scheduled_kind = parse_scheduled_kind(&kind)?;
        let result = match scheduled_kind {
            ScheduledKind::Send => {
                let payload: serde_json::Value = serde_json::from_str(&payload_json)?;
                let outcome = deliver_stored_message(
                    workspace,
                    Some(&target),
                    payload.get("content").and_then(|v| v.as_str()).unwrap_or(""),
                    None,
                    payload.get("sender").and_then(|v| v.as_str()).unwrap_or("leader"),
                    payload
                        .get("requires_ack")
                        .and_then(|v| v.as_bool())
                        .unwrap_or(false),
                    payload
                        .get("wait_visible")
                        .and_then(|v| v.as_bool())
                        .unwrap_or(false),
                    payload.get("timeout").and_then(|v| v.as_f64()).unwrap_or(30.0),
                    None,
                )?;
                serde_json::json!({
                    "ok": outcome.ok,
                    "status": status_wire(outcome.status),
                    "message_id": outcome.message_id,
                })
            }
            ScheduledKind::HealthPing => {
                event_log.write(
                    "scheduled.health_ping",
                    serde_json::json!({"event_id": id, "target": target}),
                )?;
                serde_json::json!({"ok": true, "status": "logged"})
            }
            ScheduledKind::TrustRetry => {
                let payload: serde_json::Value = serde_json::from_str(&payload_json)?;
                let outcome = handle_trust_retry_needed(
                    store,
                    &TrustRetryPayload {
                        message_id: payload
                            .get("message_id")
                            .and_then(|v| v.as_str())
                            .unwrap_or_default()
                            .to_string(),
                        attempt: payload
                            .get("attempt")
                            .and_then(|v| v.as_u64())
                            .and_then(|v| u8::try_from(v).ok())
                            .unwrap_or(1),
                        max_attempts: payload
                            .get("max_attempts")
                            .and_then(|v| v.as_u64())
                            .and_then(|v| u8::try_from(v).ok())
                            .unwrap_or(TRUST_RETRY_MAX_ATTEMPTS),
                        first_target: PaneId::new(
                            payload
                                .get("first_target")
                                .and_then(|v| v.as_str())
                                .unwrap_or(""),
                        ),
                    },
                    event_log,
                )?;
                serde_json::json!({"ok": outcome.ok, "status": status_wire(outcome.status)})
            }
        };
        let status = if result
            .get("ok")
            .and_then(serde_json::Value::as_bool)
            .unwrap_or(false)
        {
            "done"
        } else {
            "failed"
        };
        conn.execute(
            "update scheduled_events set status = ?2, fired_at = ?3, result_json = ?4 where id = ?1",
            params![id, status, chrono::Utc::now().to_rfc3339(), result.to_string()],
        )?;
        fired.push(id);
    }
    Ok(fired)
}

/// `_detect_stuck_agents` (`scheduler.py:146`):stuck 检测。**§84 守门**:`_agent_has_stuck_relevant_work`
/// 仅在有 active task / inbound message 时推送,无 pending obligation 时绝不注入探索性 prompt。
pub fn detect_stuck_agents(
    workspace: &Path,
    state: &serde_json::Value,
    store: &MessageStore,
    event_log: &EventLog,
) -> Result<Vec<String>, MessagingError> {
    let _ = (workspace, event_log);
    let team = crate::state::projection::team_state_key(state);
    let conn = crate::db::schema::open_db(store.db_path())?;
    let mut stmt = conn.prepare(
        "select agent_id, last_output_at from agent_health
         where upper(status) in ('RUNNING', 'WORKING')
           and owner_team_id = ?1
           and last_output_at is not null
           and exists (
             select 1 from messages
              where recipient = agent_health.agent_id
                and status in (
                  'pending', 'accepted', 'queued_until_idle', 'queued_until_start',
                  'queued_stopped', 'queued_pane_missing', 'target_resolved',
                  'injected', 'visible', 'submitted', 'submitted_unverified', 'delivered'
                )
           )
         order by agent_id",
    )?;
    let rows = stmt.query_map(params![team], |row| {
        Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?))
    })?;
    let mut stuck = Vec::new();
    for row in rows {
        let (agent_id, last_output_at) = row?;
        if output_is_stale(&last_output_at, 300) {
            stuck.push(agent_id);
        }
    }
    Ok(stuck)
}

fn output_is_stale(last_output_at: &str, timeout_seconds: i64) -> bool {
    let Ok(ts) = chrono::DateTime::parse_from_rfc3339(last_output_at) else {
        return false;
    };
    chrono::Utc::now()
        .signed_duration_since(ts.with_timezone(&chrono::Utc))
        .num_seconds()
        >= timeout_seconds
}

/// stuck/idle 告警列表 (`scheduler.py:222`)。team-scoped 抑制查询。CLI 调。
pub fn stuck_list(workspace: &Path) -> Result<serde_json::Value, MessagingError> {
    let state = crate::state::persist::load_runtime_state(workspace)?;
    let team = active_team_key(workspace, &state);
    let suppressed = state
        .get("coordinator")
        .and_then(|v| v.get("suppressed_idle_alerts"))
        .and_then(|v| v.get(&team))
        .cloned()
        .unwrap_or_else(|| serde_json::json!({}));
    Ok(serde_json::json!({"ok": true, "suppressed_idle_alerts": suppressed}))
}

/// 取消 idle 告警抑制 (`scheduler.py:262`)。`alert_type` 为 [`AlertType`] 或 `all` (展开全集);
/// owner-gate 校验 (check_team_owner)。CLI 调。
pub fn stuck_cancel(
    workspace: &Path,
    agent_id: &str,
    alert_type: Option<AlertType>,
    suppressed_by: &str,
) -> Result<serde_json::Value, MessagingError> {
    let alert_types: Vec<&str> = match alert_type {
        Some(AlertType::Stuck) => vec!["stuck"],
        Some(AlertType::IdleFallback) => vec!["idle_fallback"],
        Some(AlertType::CrossWorkerDeadlock) => vec!["cross_worker_deadlock"],
        None => vec!["cross_worker_deadlock", "idle_fallback", "stuck"],
    };
    let mut state = crate::state::persist::load_runtime_state(workspace)?;
    let team = active_team_key(workspace, &state);
    let now = chrono::Utc::now().to_rfc3339();
    let assigned = assigned_task_ids(&state, agent_id);
    let store = MessageStore::open(workspace)?;
    let delivered = delivered_message_ids(&store, &team, agent_id)?;
    for kind in &alert_types {
        upsert_suppression(
            &mut state,
            SuppressionRecord {
                team: &team,
                agent_id,
                alert_type: kind,
                suppressed_by,
                suppressed_at: &now,
                assigned_task_ids: assigned.clone(),
                delivered_message_ids: delivered.clone(),
            },
        );
    }
    crate::state::persist::save_runtime_state(workspace, &state)?;
    crate::event_log::EventLog::new(workspace).write(
        "coordinator.idle_alert_suppressed",
        serde_json::json!({
            "agent_id": agent_id,
            "team": team,
            "alert_types": alert_types,
            "suppressed_by": suppressed_by,
        }),
    )?;
    let suppressed = state
        .get("coordinator")
        .and_then(|v| v.get("suppressed_idle_alerts"))
        .and_then(|v| v.get(&team))
        .and_then(|v| v.get(agent_id))
        .cloned()
        .unwrap_or_else(|| serde_json::json!({}));
    Ok(serde_json::json!({
        "ok": true,
        "agent_id": agent_id,
        "alert_types": alert_types,
        "suppressed": suppressed
    }))
}

fn active_team_key(workspace: &Path, state: &serde_json::Value) -> String {
    state
        .get("active_team_key")
        .and_then(serde_json::Value::as_str)
        .filter(|team| !team.is_empty())
        .map(ToString::to_string)
        .or_else(|| workspace.file_name().map(|name| name.to_string_lossy().to_string()))
        .unwrap_or_else(|| "current".to_string())
}

fn assigned_task_ids(state: &serde_json::Value, agent_id: &str) -> Vec<String> {
    let mut ids: Vec<String> = state
        .get("tasks")
        .and_then(serde_json::Value::as_array)
        .map(|tasks| {
            tasks
                .iter()
                .filter(|task| task.get("assignee").and_then(serde_json::Value::as_str) == Some(agent_id))
                .filter_map(|task| task.get("id").and_then(serde_json::Value::as_str).map(ToString::to_string))
                .collect()
        })
        .unwrap_or_default();
    ids.sort();
    ids
}

fn delivered_message_ids(
    store: &MessageStore,
    team: &str,
    agent_id: &str,
) -> Result<Vec<String>, MessagingError> {
    let conn = crate::db::schema::open_db(store.db_path())?;
    let mut stmt = conn.prepare(
        "select message_id from messages
         where (owner_team_id = ?1 or owner_team_id is null)
           and recipient = ?2
           and status in ('visible', 'submitted', 'delivered', 'acknowledged')
         order by message_id",
    )?;
    let rows = stmt.query_map(params![team, agent_id], |row| row.get::<_, String>(0))?;
    Ok(rows.collect::<Result<Vec<_>, _>>()?)
}

struct SuppressionRecord<'a> {
    team: &'a str,
    agent_id: &'a str,
    alert_type: &'a str,
    suppressed_by: &'a str,
    suppressed_at: &'a str,
    assigned_task_ids: Vec<String>,
    delivered_message_ids: Vec<String>,
}

fn upsert_suppression(state: &mut serde_json::Value, record: SuppressionRecord<'_>) {
    let Some(root) = state.as_object_mut() else {
        return;
    };
    let Some(coordinator) = root
        .entry("coordinator")
        .or_insert_with(|| serde_json::json!({}))
        .as_object_mut()
    else {
        return;
    };
    let Some(all) = coordinator
        .entry("suppressed_idle_alerts")
        .or_insert_with(|| serde_json::json!({}))
        .as_object_mut()
    else {
        return;
    };
    let Some(team_map) = all
        .entry(record.team.to_string())
        .or_insert_with(|| serde_json::json!({}))
        .as_object_mut()
    else {
        return;
    };
    let Some(agent_map) = team_map
        .entry(record.agent_id.to_string())
        .or_insert_with(|| serde_json::json!({}))
        .as_object_mut()
    else {
        return;
    };
    agent_map.insert(
        record.alert_type.to_string(),
        serde_json::json!({
            "suppressed_at": record.suppressed_at,
            "suppressed_by": record.suppressed_by,
            "snapshot": {
                "assigned_task_ids": record.assigned_task_ids,
                "delivered_message_ids": record.delivered_message_ids
            },
        }),
    );
}
